### ISSUE TITLE
[iOS only] [v1.12+] `onFocus` event is only called on first focus on TextInput

### ISSUE THREAD

**Issue by christophemenager** [kirillzyusko/react-native-keyboard-controller#456]
**Describe the bug**
- Only on ios, `onFocus` is only called once, and not if text input is re-focused
- Same happens for the `onBlur`event
- This regression has been introduced in 1.12, 1.11 is working fine
- No issue on Android

**To Reproduce**

1. Add a TextInput centered in a View on iOS (react-native 0.73.8)
2. Click on the input, onFocus is called
3. Click outside, onBlur is called
4. Click on the input again: nothing happens (onFocus is NOT called)

**Expected behavior**
onFocus and onBlur events should be called everytime the input is focused or blurred

**Smartphone (please complete the following information):**

- Device: iPhone X
- OS: 15.x
- RN version: 0.73.8
- RN architecture: old
- JS engine: Hermes
- Library version: 1.12.1

**Additional context**
This regression has been introduced in 1.12, 1.11 is working fine

**Comment by kirillzyusko** [comment]
Hey @christophemenager 

Can you post a minimal reproduction code, please? I've used this code:

```tsx
export default function AwareScrollView({ navigation }: Props) {
  return (
    <>
      <ScrollView style={{flex: 1,}} contentContainerStyle={{paddingLeft: 100, flex: 1, justifyContent: "center", alignItems: "center", backgroundColor: "red"}}>
        <TextInput
          key={1}
          placeholder={`TextInput#${1}`}
          keyboardType={1 % 2 === 0 ? "numeric" : "default"}
          multiline={false}
          onFocus={() => console.log("onFocus")}
          onBlur={() => console.log("onBlur")}
          style={{borderWidth: 2, borderColor: "black"}}
        />
      </ScrollView>
    </>
  );
}
```

But it always calls `onFocus`/`onBlur` whenever I press on input/press outside:

```bash
 LOG  onFocus
 LOG  onBlur
 LOG  onFocus
 LOG  onBlur
 LOG  onFocus
 LOG  onBlur
 LOG  onFocus
 LOG  onBlur
```

I tested in RN 0.74 - do you think it makes a difference?

**Comment by kirillzyusko** [comment]
I also tested RN 0.73.4 (old architecture) - still everything works okay 🤷‍♂️ Guess something wrong with my code that I'm using 🤔

**Comment by kirillzyusko** [comment]
@christophemenager I've also tested on iPhone 6s (iOS 15.8) and I can not reproduce this problem in example app (I used Toolbar screen as example).

May I ask you to test my example app and maybe perform some code modifications to show how it can be reproduced?

The issues= seems to be quite critical, so I want to fix it ASAP, but I can not reproduce it 😔 

Please, don't ignore me 🙏

**Comment by christophemenager** [comment]
Sorry I am a bit busy this week, can I get back to you next Monday? I will try to give you a minimal repro code :)

**Comment by kirillzyusko** [comment]
> Sorry I am a bit busy this week, can I get back to you next Monday? 

Yeah, sure, take your time 👍 

>  I will try to give you a minimal repro code :)

Thank you ❤️

**Comment by christophemenager** [comment]
Hi there! Sorry for the delay. 

I tried to reproduce in an isolated react-native environment but I could not.

I guess it must be very specific to my setup, so I will close this issue and reopen it if I find a way to reproduce it properly.

**Comment by kirillzyusko** [comment]
@christophemenager yes, feel free to re-open this 👍 

In 1.12 version I started to inject my own delegate to intercept some events that are available only on delegate level. So I guess in some scenarios delegate substitution doesn't work properly, so it would be good if you could figure out and provide a repro 🙏

**Comment by bikowalczyk** [comment]
I'm also experiencing this issue, but only when `multiline={true}`. After the first `onBlur` no events (onFocus, onBlur, onTextChange) are called from the `TextInput`, although you can see its value changing. I was using a workaround of forcing the input to rerender (by changing its key) when the keyboard was dissmissed and it worked, but wasn't stable. 

As mentioned above reverting to `1.11.7` fixes the issue!

**Comment by kirillzyusko** [comment]
@bikowalczyk do you think it would be possible to create a simple reproduction repository?

I think it's a very critical issue but I really can't reproduce it in my example project 🤷‍♂️ Staying on `1.11.7` long time is kind of not very good way, because you'll miss a lot of great improvements.

So if you can help me and prepare a reproduction example I would highly appreciate you ❤️

**Comment by kirillzyusko** [comment]
@christophemenager yeah, the algorithm for substituting the delegate hasn't been changed since `1.12.0` version, so the issue most likely will be in newest versions as well.

Maybe you are using specific props for `TextInput`? I had a problem when layout couldn't be properly detected if devs were using `contextMenuHidden` property - maybe you are also using a combination of properties that leads to that result?

**Comment by christophemenager** [comment]
@kirillzyusko I had a bit more time to dig into this issue and here is what I found:

- It's not linked to some weird props, I reduced the number of props to its minimum and I still have the issue of `onFocus` (and `onBlur`) not being triggered after re-focus (they are triggered only once)
- The problem occurs only on iOS
- The problem was introduced in version `1.11.7` and is still present in the latest version `1.14.0`
- **It's only reproduced on iOS 15** (and maybe lower ?) but **not on iOS 16, 17 or 18**
- The problem occurs as soon as the library is added, to the project no matter if the provider is enabled or not on a given screen (I would have expect the problem to disappear if I disable the KeyBoardProvider)

I used react-native 0.75.3 for this investigation, and an iphone 13

So I guess you can easily reproduce this issue on your side following those steps:

- iPhone emulator with **ios 15**
- react-native 0.75.3
- expo 51
- `TextInput` from react-native, with only `onFocus`

I tried to repro on expo snacks but I could not find a way to select an ios version so it's not useful in this case !

Please let me know if I could help, but I guess the important thing here is that it's only on "old" versions of iOS, that's why it has never been reported a lot

**Comment by kirillzyusko** [comment]
@christophemenager I think I did everything properly:

https://github.com/user-attachments/assets/e4815c4f-e78c-4cd4-8748-37b42a15b5ee

Can you show a video how it looks in your case?

The code is:

```jsx
import "react-native-gesture-handler";

import * as React from "react";
import { ActivityIndicator, StatusBar, StyleSheet, TextInput } from "react-native";
import { GestureHandlerRootView } from "react-native-gesture-handler";
import { KeyboardProvider } from "react-native-keyboard-controller";
import {
  SafeAreaProvider,
  initialWindowMetrics,
} from "react-native-safe-area-context";

const styles = StyleSheet.create({
  root: {
    flex: 1,
  },
});

export default function App() {
  return (
    <SafeAreaProvider initialMetrics={initialWindowMetrics}>
      <GestureHandlerRootView style={styles.root}>
        <KeyboardProvider statusBarTranslucent>
          <TextInput onFocus={() => console.log("onFocus")} style={{ width: 200, height: 50, backgroundColor: "yellow", marginTop: 50 }} />
        </KeyboardProvider>
      </GestureHandlerRootView>
    </SafeAreaProvider>
  );
}
```

Setup that I used:
- iPhone 13 **Pro** (iOS 15.5);
- react-native 0.75.3
- it's **not** Expo project (it's bare RN)
- `TextInput` from `react-native`, with only `onFocus` and `style`

> The problem was introduced in version 1.11.7

Are you sure it was introduced in `1.11.7` and not `1.12.0`? Because the issue has a different title 🙈 

Any ideas what is the difference between our setups and why it doesn't work on your machine, but works on mine? The only difference that I see is that you are using Expo and I'm using pure RN project. Do you think it can be a key difference?

**Comment by christophemenager** [comment]
@kirillzyusko thanks for your quick reply. 

My mistake, this issue was introduced in 1.12.0, I edited my last comment.

I will keep digging and send you more info later today.

**Comment by kirillzyusko** [comment]
@christophemenager thank you! Hope we eventually can figure out what's exactly happening there 👀

**Comment by christophemenager** [comment]
It seems I managed to repro it on a simple expo example! 🎉

https://github.com/user-attachments/assets/c79e0b28-dee7-4b39-b4e2-e64a9fa52dd4

```
  "dependencies": {
    "@expo/vector-icons": "^14.0.2",
    "@react-navigation/native": "^6.0.2",
    "expo": "~51.0.28",
    "expo-constants": "~16.0.2",
    "expo-font": "~12.0.9",
    "expo-linking": "~6.3.1",
    "expo-router": "~3.5.23",
    "expo-splash-screen": "~0.27.5",
    "expo-status-bar": "~1.12.1",
    "expo-system-ui": "~3.0.7",
    "expo-web-browser": "~13.0.3",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-native": "0.74.5",
    "react-native-gesture-handler": "~2.16.1",
    "react-native-keyboard-controller": "^1.14.0",
    "react-native-library": "^1.0.25",
    "react-native-reanimated": "~3.10.1",
    "react-native-safe-area-context": "4.10.5",
    "react-native-screens": "3.31.1",
    "react-native-web": "~0.19.10"
  },
```
Extra infos:
- Iphone 13, ios 15.2
- Something I discovered while playing with my app on emulator: the problem only occurs when the keyboard opens. On emulator, if you don't trigger (CMD+K) the keyboard on focus, then the problem is not visible

**Comment by kirillzyusko** [comment]
@christophemenager would you mind to publish reproduction example? And maybe you have a discord or any other messenger so that we can chat with you (I think it'll be more productive)?

> if you don't trigger (CMD+K) the keyboard on focus, then the problem is not visible

So it's visible only if you trigger a keyboard via CMD+K, right? Can you say when exactly you press these keys, i. e. set focus, keyboard is not shown and then press CMD+K (according to the video)?

Have you tried to reproduce it on a real device?

**Comment by kirillzyusko** [comment]
I could reproduce it only one time - was super happy, but when I modified code the issue is not reproducible anymore 🤯 

My brain is 🤯

**Comment by christophemenager** [comment]
> @christophemenager would you mind to publish reproduction example? And maybe you have a discord or any other messenger so that we can chat with you (I think it'll be more productive)?

Yes and yes, but not this week, I will get back to you next week :) 

> So it's visible only if you trigger a keyboard via CMD+K, right? Can you say when exactly you press these keys, i. e. set focus, keyboard is not shown and then press CMD+K (according to the video)?

On the video record, If I remember well, I am not pressing CMD+K, I only click on the input. I made other tests in other conditions and I noticed that I could re-focus without any issue while the keyboard do not open. As soon as it opens, the onFocus is not called again. Maybe it's a coincidence, but I thought it could be a hint for you to know where to start :)

**Comment by kirillzyusko** [comment]
> On the video record, If I remember well, I am not pressing CMD+K, I only click on the input. I made other tests in other conditions and I noticed that I could re-focus without any issue while the keyboard do not open. As soon as it opens, the onFocus is not called again. Maybe it's a coincidence, but I thought it could be a hint for you to know where to start :)

Well, will try to do more testing, but at the moment it doesn't give me any hints 🙃

**Comment by christophemenager** [comment]
@kirillzyusko I will try to push a repro repo today :)

**Comment by christophemenager** [comment]
@kirillzyusko well I tried to reproduce again this morning and... I failed 🤦 Trying around 50 times I managed to reproduce only one time, but I could not find how :/ 

I tried to tackle this issue from the other way: trying to simplify my project (where I can systematically reproduce) until I solve the issue so I could find what is exactly is causing the issue in my current project setup. 
I cleaned every component and had up to having a simple textInput in a simple react-native scrollView, but I still had the issue.

My feeling is that it could come from an other dependency installed in my project that messes up with yours, but I am really not sure.

The only thing I know for sure is that in my project, I reproduced on iOS 15.2 but not on 16,17 and 18.

As 
- we both tried to reproduce but could not manage
- the issue seems pretty specific to my setup and limited to "old" ios versions

I suggest we close the issue for now and wait for more inputs if others have the same problem.

**Comment by kirillzyusko** [comment]
@christophemenager okay, I don't see anything suspicious 🙈 If you can create an empty project, add all these dependencies and the issue will be reproducible - let me know and I'll have a look on a reproduction example.

As you suggested - I'll close the issue, but if you (or anyone else) find anything new to share, feel free to do it here and I'll be happy to re-open that issue!

**Comment by christophemenager** [comment]
@kirillzyusko ahah such a warrior fighting again this bug, I admire that ;) Sure, can you just give me the path of the file where I should add this piece of code?

**Comment by kirillzyusko** [comment]
@christophemenager sure, it's located here:

https://github.com/kirillzyusko/react-native-keyboard-controller/blob/56cd863e90c7b0dc99541aeccc657ca12fa4805b/ios/observers/FocusedInputObserver.swift#L206-L212

You need to replace this function with a code that I provided 🙂

**Comment by christophemenager** [comment]
@kirillzyusko good news, it fixes the issue !! 🎉

|Before|After|
|---|---|
|https://github.com/user-attachments/assets/a12dcf98-b3ea-4d6f-9c02-c0d132c4e3ac|https://github.com/user-attachments/assets/06b1ed57-aa11-4e13-b11d-8233aa5a0bdf|

We can't see it but on the "before" video I am pressing multiple times on the first inputs but they are not getting focused.

I made sure to revert the change and I reproduced the issue again so I confirm this piece of codes you gave me fixes the issue !

**Comment by kirillzyusko** [comment]
Amazing @christophemenager ! 🎉 I'll submit a PR with a fix today then and will publish a fix under `1.14.2` version 👀 

May I also ask you to test this variant of the code?

```swift
  private func substituteDelegateBack(_ input: UIResponder?) {
    if let textField = input as? UITextField, let oldDelegate = delegate.activeDelegate as? UITextFieldDelegate {
      textField.delegate = oldDelegate
    } else if let textView = input as? UITextView, let oldDelegate = delegate.activeDelegate as? UITextViewDelegate {
      (textView as? RCTUITextView)?.setForceDelegate(oldDelegate)
    }
  }
```

Just want to understand better what's going wrong internally to apply a better suitable fix 🙃

**Comment by christophemenager** [comment]
> May I also ask you to test this variant of the code?

It fixes the issue too :)
